File: extensions/vscode/src/commands/override.ts
import * as path from "node:path";
import * as vscode from "vscode";
import type { BinaryManager } from "../binary/manager";
import type { CliClient } from "../cli/client";
import { executeAsTask } from "../providers/taskProvider";

const TOKEN_HELP =
  "Tokens: @cmd.sub  +channel  KEY=val  /test-args  --flags  |  @ append  !! clear  !env";

/**
 * Cmd+Shift+R — prompt for override tokens at cursor.
 */
export async function overrideAtCursor(
  binaryManager: BinaryManager,
  client: CliClient,
  output: vscode.OutputChannel,
): Promise<void> {
  const editor = vscode.window.activeTextEditor;
  if (!editor || editor.document.languageId !== "rust") {
    vscode.window.showErrorMessage("Cargo Runner: open a Rust file first");
    return;
  }

  const filePath = editor.document.uri.fsPath;
  const line = editor.selection.active.line + 1;
  await overrideAt(binaryManager, client, output, `${filePath}:${line}`);
}

/** Override for an explicit file:line (CodeLens / tree). */
export async function overrideAt(
  binaryManager: BinaryManager,
  client: CliClient,
  output: vscode.OutputChannel,
  fileArg: string,
  prefill?: string,
): Promise<void> {
  const filePath = fileArg.replace(/:\d+$/, "");
  const lineMatch = fileArg.match(/:(\d+)$/);
  const line = lineMatch ? lineMatch[1] : "1";
  const cwd =
    vscode.workspace.getWorkspaceFolder(vscode.Uri.file(filePath))?.uri
      .fsPath || path.dirname(filePath);

  // Navigate to target so user sees context
  try {
    const doc = await vscode.workspace.openTextDocument(filePath);
    const editor = await vscode.window.showTextDocument(doc);
    const line0 = Math.max(0, parseInt(line, 10) - 1);
    const pos = new vscode.Position(line0, 0);
    editor.selection = new vscode.Selection(pos, pos);
  } catch {
    // file may not open; still try override
  }

  const tokensInput = await vscode.window.showInputBox({
    title: "Cargo Runner Override",
    prompt: TOKEN_HELP,
    placeHolder: "e.g. @dx.serve --release RUST_LOG=debug /--nocapture",
    value: prefill || "",
    ignoreFocusOut: true,
    validateInput: (value) => {
      if (value.trim() === "@.") {
        return "Incomplete @cmd.sub token";
      }
      return null;
    },
  });

  if (tokensInput === undefined) {
    return;
  }

  const tokens = tokenize(tokensInput.trim());
  const binary = await binaryManager.ensureBinary();

  try {
    const dry = await client.dryRun(fileArg, cwd);
    output.appendLine(`Before override: ${dry.shell}`);
  } catch (e) {
    output.appendLine(`Dry-run before override: ${e}`);
  }

  if (tokens.length === 0) {
    await executeAsTask(binary, ["run", fileArg], {
      cwd,
      label: `Cargo Runner: ${path.basename(filePath)}:${line}`,
    });
    return;
  }

  if (tokens.length === 1 && (tokens[0] === "-" || tokens[0] === "!!")) {
    await client.setOverride(fileArg, tokens, cwd);
    vscode.window.showInformationMessage("Cargo Runner: override removed");
    return;
  }

  const action = await vscode.window.showQuickPick(
    [
      {
        label: "$(save) Save & Run",
        description: "Write override to .cargo-runner.json then run",
        value: "save-run",
      },
      {
        label: "$(eye) Preview only",
        description: "Save override and show dry-run JSON",
        value: "preview",
      },
      {
        label: "$(check) Save only",
        description: "Write override without running",
        value: "save",
      },
    ],
    { placeHolder: "What should we do with this override?" },
  );

  if (!action) {
    return;
  }

  try {
    await client.setOverride(fileArg, tokens, cwd);
    output.appendLine(`Override saved: ${tokens.join(" ")}`);
  } catch (e) {
    vscode.window.showErrorMessage(`Failed to save override: ${e}`);
    return;
  }

  if (action.value === "save") {
    vscode.window.showInformationMessage("Cargo Runner: override saved");
    return;
  }

  if (action.value === "preview") {
    try {
      const dry = await client.dryRun(fileArg, cwd);
      output.appendLine(`After override: ${JSON.stringify(dry, null, 2)}`);
      output.show(true);
      vscode.window.showInformationMessage(`Preview: ${dry.shell}`);
    } catch (e) {
      vscode.window.showErrorMessage(`Preview failed: ${e}`);
    }
    return;
  }

  await executeAsTask(binary, ["run", fileArg], {
    cwd,
    label: `Cargo Runner: ${path.basename(filePath)}:${line}`,
  });
}

/** Split like a shell while keeping KEY=value and quoted strings. */
export function tokenize(input: string): string[] {
  if (!input) {
    return [];
  }
  return (
    input
      .match(/(?:[^\s"]+|"[^"]*")+/g)
      ?.map((t) => t.replace(/^"|"$/g, "")) ?? []
  );
}

File: crates/cli/src/assets/nvim-plugin/lua/cargo_runner/override.lua
local cli = require("cargo_runner.cli")
local job = require("cargo_runner.job")
local notify = require("cargo_runner.notify")

local M = {}

local TOKEN_HELP =
  "Tokens: @cmd.sub  +channel  KEY=val  /test-args  --flags  |  @ append  !! clear  !env"

---Simple shell-ish tokenize (handles quotes loosely).
---@param s string
---@return string[]
function M.tokenize(s)
  local tokens = {}
  local i = 1
  local n = #s
  while i <= n do
    while i <= n and s:sub(i, i):match("%s") do
      i = i + 1
    end
    if i > n then
      break
    end
    local c = s:sub(i, i)
    if c == '"' or c == "'" then
      local q = c
      i = i + 1
      local start = i
      while i <= n and s:sub(i, i) ~= q do
        i = i + 1
      end
      table.insert(tokens, s:sub(start, i - 1))
      if i <= n then
        i = i + 1
      end
    else
      local start = i
      while i <= n and not s:sub(i, i):match("%s") do
        i = i + 1
      end
      table.insert(tokens, s:sub(start, i - 1))
    end
  end
  return tokens
end

local ACTIONS = {
  ["Save & Run"] = "save-run",
  ["Preview only (dry-run)"] = "preview",
  ["Save only"] = "save",
}

local function after_save_actions(file_arg, cwd, tokens)
  vim.ui.select({
    "Save & Run",
    "Preview only (dry-run)",
    "Save only",
  }, {
    prompt = "Cargo Runner override — next step",
  }, function(choice)
    if not choice then
      return
    end
    local value = ACTIONS[choice]
    if not value then
      return
    end

    local ok, err = pcall(cli.set_override, file_arg, tokens, cwd)
    if not ok then
      notify.error("Failed to save override: " .. tostring(err))
      return
    end

    if value == "save" then
      notify.success("Override saved")
      return
    end

    if value == "preview" then
      local dok, dry = pcall(cli.dry_run, file_arg, cwd)
      if not dok then
        notify.error("Preview failed: " .. tostring(dry))
        return
      end
      notify.info("Preview: " .. tostring(dry.shell))
      require("cargo_runner.ui.error_float").open(
        "After override:\n" .. (dry.shell or vim.inspect(dry)),
        { title = "Cargo Runner — override preview" }
      )
      return
    end

    -- save-run
    notify.success("Override saved — running")
    local label = vim.fn.fnamemodify(file_arg:gsub(":%d+$", ""), ":t")
      .. ":"
      .. (file_arg:match(":(%d+)$") or "?")
    job.run({ file_arg = file_arg, cwd = cwd, label = label })
  end)
end

function M.at_cursor()
  local file_arg, err = cli.cursor_file_arg()
  if not file_arg then
    notify.error(err or "cannot override")
    return
  end
  local cwd = cli.project_cwd()

  -- Optional dry-run before for context
  local before = nil
  local dok, dry = pcall(cli.dry_run, file_arg, cwd)
  if dok and dry and dry.shell then
    before = dry.shell
  end

  local prompt = TOKEN_HELP
  if before then
    prompt = "Current: " .. before .. "\n" .. TOKEN_HELP
  end

  vim.ui.input({
    prompt = "Cargo Runner Override: ",
    default = "",
  }, function(input)
    if input == nil then
      return -- cancelled
    end

    local trimmed = vim.trim(input)
    local tokens = M.tokenize(trimmed)

    if #tokens == 0 then
      -- empty → run only (VS Code behavior)
      local label = vim.fn.fnamemodify(file_arg:gsub(":%d+$", ""), ":t")
        .. ":"
        .. (file_arg:match(":(%d+)$") or "?")
      job.run({ file_arg = file_arg, cwd = cwd, label = label })
      return
    end

    if #tokens == 1 and (tokens[1] == "-" or tokens[1] == "!!") then
      local ok, oerr = pcall(cli.set_override, file_arg, tokens, cwd)
      if not ok then
        notify.error("Failed to remove override: " .. tostring(oerr))
        return
      end
      notify.success("Override removed")
      return
    end

    if trimmed == "@." then
      notify.error("Incomplete @cmd.sub token")
      return
    end

    -- Show help in echo if useful
    if before then
      vim.api.nvim_echo({ { prompt, "Comment" } }, false, {})
    end

    after_save_actions(file_arg, cwd, tokens)
  end)
end

return M
